akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1195,10 +1195,12 @@ private[stream] object Collect {
           future.value match {
             case None ⇒ future.onComplete(holder)(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
             case Some(v) ⇒
-              // #20217 the future is already here, avoid scheduling it on the dispatcher
+              // #20217 the future is already here, optimization: avoid scheduling it on the dispatcher and
+              // run the logic directly on this thread
               holder.setElem(v)
               v match {
-                case Failure(ex) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop ⇒ failStage(ex)
+                // this optimization also requires us to stop the stage to fail fast if the decider says so:
+                case Failure(NonFatal(ex)) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop ⇒ failStage(ex)
```

Comment:
I don't think adding `NonFatal` is necessary here. We don't do that anywhere and afair this is taken care of in the future implementation.

Reply:
Might be true, but was already like that in the other places in this stage. Looked at `Future` but can't see that it would filter fatal errors anywhere.

Reply:
I think it can only happen if someone calls `Promise.failure(fatalError)` manually. And in that case just ignoring the error doesn't seem like the right thing to do either... ;)

Reply:
Fixed